dask/highlevelgraph.py

```diff
@@ -515,6 +515,216 @@ def layer_info_dict(self):
         return info
 
 
+class AbstractLayer(Layer):
```

Comment:
Could this be just `Layer`?  Or does this conflict somehow with `MaterializedLayer`?

Reply:
I'll need to think a bit more about the best structure here, but my original assumption was that we want to define three general classes in `highlevelgraph.py`: A base `Layer` class (defining the required methods), a special subclass for "materialized" layers, and a special subclass for "non-materialized" layers.  This PR is essentially taking most of the existing layer definitions and moving them under the same "non-materialized" (`AbstractLayer`) umbrella. The follow-up plan was to make sure everything is either a `MaterializedLayer` or an `AbstractLayer`.

It may indeed be possible to avoid a distinct `AbstractLayer` class by implementing the same logic in `Layer`. The main reason I didn't attempt to do this was that my `AbstractLayer` definition was making a lot of assumptions about how the layer should implement culling, and I was hoping to leave the porting of `ArrayOverlapLayer` for follow-up work.  However, I will likely try to explore the `Layer`/`MaterializedLayer` design today/tomorrow since it would probably be more intuitive.